### ISSUE TITLE
Training with multiple TPU cores is incorrect and slower

### ISSUE THREAD

**Issue by nalzok** [huggingface/accelerate#450]
### System Info

```Shell
- `Accelerate` version: 0.10.0
- Platform: Linux-5.13.0-1023-gcp-x86_64-with-glibc2.29
- Python version: 3.8.10
- Numpy version: 1.22.4
- PyTorch version (GPU?): 1.11.0a0+gitbc2c6ed (False)
- `Accelerate` default config:
        - compute_environment: LOCAL_MACHINE
        - distributed_type: TPU
        - mixed_precision: no
        - use_cpu: False
        - num_processes: 8
        - machine_rank: 0
        - num_machines: 1
        - main_process_ip: None
        - main_process_port: None
        - main_training_function: main
        - deepspeed_config: {}
        - fsdp_config: {}
```

### Information

- [X] The official example scripts
- [ ] My own modified scripts

### Tasks

- [X] One of the scripts in the examples/ folder of Accelerate or an officially supported `no_trainer` script in the `examples` folder of the `transformers` repo (such as `run_no_trainer_glue.py`)
- [ ] My own task or dataset (give details below)

### Reproduction

First, train with 8 TPU cores (run the script twice and count the second one to exclude the time used to download the datasets)

```python3
$ pipenv run accelerate config
In which compute environment are you running? ([0] This machine, [1] AWS (Amazon SageMaker)): 0
Which type of machine are you using? ([0] No distributed training, [1] multi-CPU, [2] multi-GPU, [3] TPU): 3
What is the name of the function in your script that should be launched in all parallel scripts? [main]:
How many TPU cores should be used for distributed training? [1]:8

$ time pipenv run accelerate launch accelerate/examples/nlp_example.py
...a lot of irrelevant warnings...
epoch 0: {'accuracy': 0.3125, 'f1': 0.0}
epoch 1: {'accuracy': 0.671875, 'f1': 0.8004750593824228}
epoch 2: {'accuracy': 0.6796875, 'f1': 0.8079625292740047}

________________________________________________________
Executed in  157.17 secs   fish           external
   usr time  474.51 secs  592.00 micros  474.51 secs
   sys time   99.76 secs    0.00 micros   99.76 secs
```

Next, train with 1 TPU core

```python
$ pipenv run accelerate config
In which compute environment are you running? ([0] This machine, [1] AWS (Amazon SageMaker)): 0
Which type of machine are you using? ([0] No distributed training, [1] multi-CPU, [2] multi-GPU, [3] TPU): 3
What is the name of the function in your script that should be launched in all parallel scripts? [main]:
How many TPU cores should be used for distributed training? [1]:1

$ time pipenv run accelerate launch accelerate/examples/nlp_example.py
...a lot of irrelevant warnings...
epoch 0: {'accuracy': 0.6838235294117647, 'f1': 0.8122270742358079}
epoch 1: {'accuracy': 0.7205882352941176, 'f1': 0.8256880733944955}
epoch 2: {'accuracy': 0.7156862745098039, 'f1': 0.8215384615384616}

________________________________________________________
Executed in  150.81 secs   fish           external
   usr time  352.98 secs  419.00 micros  352.98 secs
   sys time   34.38 secs  210.00 micros   34.38 secs
```

### Expected behavior

```Shell
+ The `accuracy` and `f1` should be very close, but they are way off, and things are particularly weird in epoch 0 when we use 8 cores.
+ Using 8 TPU cores should be much faster, but it's slower (157.17s vs 150.81s).
```

**Comment by sgugger** [comment]
This is not a bug, you are multiplying the batch size by 8 when training on TPU, so it's logical to get different results. You probably need to adapt the other hyperparameters a bit to get the same results.

**Comment by muellerzr** [comment]
@nalzok as Sylvain recommended, to be able to compare these properly we need to multiply the bs by 8. Here's the results after I've done so, which shows they're quite similar. 

Re; speed, my best guess is it it might be due to communicating on all 8 cores + rendezvous etc. We also do not train very long, and the speedup should be faster when training for longer periods of time I believe. 

From main:
![image](https://user-images.githubusercontent.com/7831895/175563001-d38ba39d-19dd-4931-90cf-5094f31744f3.png)

On Accelerate 0.10.0 (which I'd consider close enough):
![image](https://user-images.githubusercontent.com/7831895/175566253-3b16a3cd-191b-41a5-af75-d486174531cd.png)

**Comment by nalzok** [comment]
Interesting. I just tried reproducing the bug with a fresh install of `accelerate`, and I am no longer getting the weird `epoch 0: {'accuracy': 0.3125, 'f1': 0.0}`. I guess that means the bug has been fixed, or maybe it was only a little glitch on my local system.

I want to keep the issue open for a little longer though. Could you provide a simple example where using 8 TPU cores is significantly faster than using 1 core? I understand there is synchronization overhead, so I would accept even a 2x speed up with 8x more hardware resources, but using more hardware really shouldn't make the computation slower.

**Comment by nalzok** [comment]
Thanks for your time!

To clarify, I am on a TPU v3-8 machine on GCP (obtained through [TRC](https://sites.research.google/trc/about/)), not Colab. Additionally, the architecture is a TPU VM instead of a TPU node, so there are no networking roundtrips (i.e. `$XRT_TPU_CONFIG` is `localservice;0;localhost:51011`).

**Comment by muellerzr** [comment]
@nalzok I can stem the issue to the fact that the batch size on the eval vs train dataloaders are different, causing memory reallocation to occur. Can you try making the train and eval bs to be the same? (still working on getting myself setup).

But doing this in colab I got the following times:

- 8 cores first time: 152s
- 8 cores second run: 74.74s
- 1 core w/ 8x batch size to keep equal: 125s

**Comment by muellerzr** [comment]
@nalzok the bug is indeed with the NLP example script, I'll put in a fix tommorow. What I did here was took the NLP Example notebook (found [here](https://github.com/huggingface/notebooks/blob/master/examples/accelerate/simple_nlp_example.ipynb) and I could get the right training results on 8 cores
![image](https://user-images.githubusercontent.com/7831895/176052283-fd88e45a-e88d-48c7-80bb-c78fd91047a2.png)
